Playing functions

itivoli opened the pull request:
The `AudioGenerator` is now more node like and connects to the `LowerTimeline` using a signal that indicates if audio playback is finished from the timeline's end. I've updated `Toolbar.cs` to interface with the timeline. The most important change is that constant functions and rational functions can now be sonified. The Timeline UI is not yet fully available, so I've interfaced the `AudioGenerator` directly with the toolbar and so currently when the play button is pressed in the application, the LowerTimeline will play all functions statically added to it via the `AddTests()` method. 

To complete the goal of playing function audio over different ranges, provisions should be made in either the `FunctionPalette` where an individual function's bounds can be modified as the user decides. In furtherance of this, I've updated the `StartTime` and `EndTime` properties of the `Function` node and they can be directly modified to update the functions bounds. So whereas the Function Palette currently looks like this:
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/fc48e7f8-6db1-4f07-85cc-521e43df8362)

the described change and associated functionality could look like this:
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/be4dd276-e2d5-464b-ad58-ea76963cfb8c)

To be next worked on: pausing the timeline.

MW-0317 committed:
Replaced IFunctionAST with Function as the current selected function representation

itivoli committed:
Merge branch 'FP-AST-TO-FUNCTION' into Playing-Functions

itivoli committed:
Now Supported: Constant Functions, Rational Functions. Toolbar Updates.

Audio playback has been updated to support constant functions and rational functions. The Toolbar (UI) has been updated to interface with the Timeline (specifically to Play the timeline via the Timeline UI). Timeline (UI) has also been renamed to 'UpperTimeline' (UI) in order to reflect its connection to the associated LowerTimeline (Sonifcation).

PROMETHIA-27 commented on Sonifcation/Function.cs:
```diff
@@ -19,15 +21,45 @@ public partial class Function : Node {
+	/// <summary>
+	/// If <c> true </c>, Function is a constant function (i.e., y = 1, y = 3, y = 6942).
+	/// </summary>
+	public bool IsConstant {get; set;}
```
Why do you need to explicitly handle the case where the function is constant?

itivoli replied:
```
// Find the full range of values that the function can take.
		double minVal = FunctionTable.Min();
		double maxVal = FunctionTable.Max();
		double functionRange = maxVal - minVal;
		
		// Map the Functions values to note numbers
		foreach(double value in FunctionTable) {
			int noteNumber;
			if(IsConstant) noteNumber = (int) Math.Abs(FunctionTable[0]) % noteRange;
			else {
				double normalizedValue = (value - minVal) / functionRange;
				noteNumber = noteNumStart + (int) (normalizedValue * noteRange);
			}
			noteSequence.Add(noteNumber);
		}
```

So, looking at `functionRange`, this is the allowable range of function outputs calculated as the difference between the maximum output and maximum input from when they were calculated using `FillFunctionTable()`. In a constant function, this ends up being 0 and that carries into the calculation for `noteNumber` and makes it infinity thus throwing the access to `Notes` in `Frequencies.cs` that holds the notes to play out of bounds.

PROMETHIA-27 commented on Sonifcation/Function.cs:
```diff
@@ -107,34 +139,66 @@ public partial class Function : Node {
+			int noteNumber;
+			if(IsConstant) noteNumber = (int) Math.Abs(FunctionTable[0]) % noteRange;
+			else {
+				double normalizedValue = (value - minVal) / functionRange;
```
Normalizing the function range doesn't seem like the way to go; this way `x + 3` and `x + 5` would sound exactly the same even though the latter is higher than the former. Unless I'm mistaken on how this works?

PROMETHIA-27 replied:
I would expect it to map the output of the function to the note range in the same way it maps constant functions now. Then you can just stop handling constant functions explicitly at all

itivoli replied:
They will sound the same, that is the primary limitation of doing it in this way. If I used a modified version of how I handled the constant functions, it would look something like `noteNumber = (int) (Math.Abs(value) + Frequencies.tuningKeyNum) % noteRange;` (provided I make `tuningKeyNum` public). Doing this would create a Note Space for the function that looks like and sounds like this:
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/9563f6dc-69cb-4b96-8159-751633bd8e46)

as opposed to looking and sounding closer to this when using a linear mapping:
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/ca830b9d-68ee-42c8-b3e9-4d56e756f12a)

Your suggestion is what I originally had in mind. I stopped trying to work it that way because the notes were just too discrete and I moved to try and get a functional audio interface using the linear mapping instead with the idea being that every function sounds differently with respect to its own space. A way to mitigate the limitation might be to add a small but significant bias to similar functions with identical ranges (as this limitation only shows up in this way) and by doing so make sure their output isn't identical. This is a band aid solution though and I am willing to go back and reevaluate how I could make the original idea work and find a different way to calculate the proper audio output.

PROMETHIA-27 replied:
Why not just increase the vertical resolution with the constant method? You can use `%` with floats to directly map any sample into a reasonable range without clamping it to integer values, so you'd still get a smooth mapping as with linear instead of discrete steps.

itivoli replied:
That was part of the original idea I ended up scrapping. I think the way you described it is the same way I had it implemented while fleshing things out in the beginning. Doing that would result in modifying the `FillNoteSequence()` method to look something like this:

```
double startFreq = Frequencies.Notes[Frequencies.startingNoteNumber];
double endFreq = Frequencies.Notes[Frequencies.endingNoteNumber];
noteNumber = (Math.Abs(value) + (int) Frequencies.tuningFreq + 1) % (endFreq) + startFreq;
```

This takes the same kind of logic I used to play constant functions and applies it to a frequency space stripped of notes. Here `noteNumber` is basically just the frequency to generate audio for and you would adjust the 'GenerateNoteAudio()' method accordingly to accommodate for this change in meaning.  `noteNumber` is also biased by translating the Functions f(t) values by adding to them `Frequencies.tuningFreq` so that calculated frequencies can guarantee that a meaningfully audible noise will be played (if I didn't do this, frequencies calculated in this manner would be low and borderline inaudible due to how tiny f(t) might be  function. This scheme produces a frequency space like so, of course linear like the direct linear mapping technique: 
![image](https://github.com/F-nk-Sound/SoundOfFunctions/assets/139672232/dc89dbe2-22d3-4dbb-af9a-74503ad79d27)

The problem with this approach lies in that the range of sound played this way is so small that the difference in tones as we transition from beginning to end is just barely noticeable. The example graph above shows only a difference of 10 Hz between the initial and final notes played and it doesn't get much better for other functions unless you have a much larger range in the time domain. If you want to try it out and see how it sounds, I've updated the branch with this method and you'd just have to load it and press the play button on the toolbar. Further explained in the most recent commit is how to switch between methods of playing the audio if you want to compare all three we've discussed (Direct Linear Mapping, Mapping Directly to the Notes and Rounding, Mapping Directly to Frequency Space w/ No Rounding).

PROMETHIA-27 replied:
I was thinking more like this (pseudocode):
`noteNumber = (Function.At(t) + C5) % 88.0;`
This way 0 would always be C5, every 1 y value goes up 1 note or down 1 note, looping back to the other end of the scale at the ends. Rather than grabbing the min and max notes and assigning those as the frequency space of the function.

itivoli replied:
Funnily enough, that's pretty much the same thing I wrote in case 2 of the Function class `FillNoteSequence()` method but a more basic version:
```
case 2:
	double note = (Math.Abs(value) + Frequencies.tuningKeyNum) % noteRange + Frequencies.startingNoteNumber;
	// Round up if necessary.
	if(Math.Ceiling(note) - note <= 0.5) noteNumber = Math.Ceiling(note);
	else noteNumber = Math.Floor(note);
	break;
```

The difference between this version and the pseudocode is that this takes into account the possibility of starting form different notes other than 1. Using this with the piano notes scheme (88 discrete notes mapped to frequencies), we have to round up or truncate that results. Kind of talked about that earlier in this thread if you want to read more about it again and why I thought we might need to find a better way to sonify.

PROMETHIA-27 commented on Sonifcation/LowerTimeline.cs:
```diff
@@ -12,201 +12,218 @@ namespace Sonification;
+	/// Provides audio playback via <c> LowerTimeline.player </c>.
+	/// </summary>
+	/// <exception cref="NotImplementedException"></exception>
+	public void Sonfiy() {
```
typo
```suggestion
	public void Sonify() {
```

PROMETHIA-27 commented on ui.gd:
Is this file needed anymore?

itivoli replied:
I left it in because I wasn't sure if Christian or Mark might need it later for whatever reason but there's no functional use of it as of now.

PROMETHIA-27 replied:
Might as well delete it, won't be hard to add it back if it's required

itivoli replied:
will do

itivoli committed:
Update Sonifcation/LowerTimeline.cs

Co-authored-by: PROMETHIA-27 <42193387+PROMETHIA-27@users.noreply.github.com>

itivoli committed:
Delete ui.gd

itivoli committed:
Added 2 extra methods of Audio Generation. Fixed Audio Cracking.

2 new methods of generating audio have been added for testing purposes. They are accessible by modifying `AudioDebugging.Method` to be 1, 2, or 3. Method 1 is the Direct Linear mapping technique. Method 2 is a modified version of Method 1 where there is no normalization and notes numbers are calculated with respect to the `noteRange` and rounded up or down accordingly to fit within the available notes. Method 3 drops notes as a form of audio playing and interacts directly with the frequency space in recognition that discrete notes aren't needed to play audio.

Also, slight audio cracking audible between notes was fixed, and a new way of handling constant functions that takes into account constant functions that might be below the programs starting note number visble in `Frequencies.cs`.
